src/components/NewBranchDialog.tsx

```diff
@@ -0,0 +1,443 @@
+        className={classes(listItemClass, isBase ? activeListItemClass : null)}
+        key={branch.name}
+        onClick={this._onBranchClickFactory(branch.name)}
+      >
+        <span className={classes(listItemIconClass, 'jp-Icon-16')} />
+        <div className={listItemContentClass}>
+          <p
+            className={classes(
+              listItemTitleClass,
+              isBold ? listItemBoldTitleClass : null
+            )}
+          >
+            {branch.name}
+          </p>
+          {desc ? <p className={listItemDescClass}>{desc}</p> : null}
+        </div>
+      </ListItem>
+    );
+  }
+
+  /**
+   * Adds model listeners.
+   */
+  private _addListeners(): void {
+    // When the HEAD changes, decent probability that we've switched branches:
+    this.props.model.headChanged.connect(this._syncState, this);
+
+    // When the status changes, we may have checked out a new branch (e.g., via the command-line and not via the extension) or changed repositories:
+    this.props.model.statusChanged.connect(this._syncState, this);
+  }
+
+  /**
+   * Removes model listeners.
+   */
+  private _removeListeners(): void {
+    this.props.model.headChanged.disconnect(this._syncState, this);
+    this.props.model.statusChanged.disconnect(this._syncState, this);
+  }
+
+  /**
+   * Syncs the component state with the underlying model.
+   */
+  private _syncState(): void {
+    const repo = this.props.model.pathRepository;
+    this.setState({
+      base: repo ? this.state.base : '',
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    });
+  }
+
+  /**
+   * Callback invoked upon closing the dialog.
+   *
+   * @param event - event object
+   */
+  private _onClose = (): void => {
+    this.props.onClose();
+    this.setState({
+      name: '',
+      filter: ''
+    });
+  };
+
+  /**
+   * Callback invoked upon a change to the menu filter.
+   *
+   * @param event - event object
+   */
+  private _onFilterChange = (event: any): void => {
+    this.setState({
+      filter: event.target.value
+    });
+  };
+
+  /**
+   * Callback invoked to reset the menu filter.
+   */
+  private _resetFilter = (): void => {
+    this.setState({
+      filter: ''
+    });
+  };
+
+  /**
+   * Returns a callback which is invoked upon clicking a branch name.
+   *
+   * @param branch - branch name
+   * @returns callback
+   */
+  private _onBranchClickFactory(branch: string) {
+    const self = this;
+    return onClick;
+
+    /**
+     * Callback invoked upon clicking a branch name.
+     *
+     * @private
+     * @param event - event object
+     */
+    function onClick() {
+      self.setState({
+        base: branch
+      });
+    }
+  }
+
+  /**
+   * Callback invoked upon a change to the branch name input element.
+   *
+   * @param event - event object
+   */
+  private _onNameChange = (event: any): void => {
+    this.setState({
+      name: event.target.value
```

Comment:
@telamonian On my MacBook, I am unable to hold down a key to generate repeats. Even within this comment box on GitHub, I have to press each time I want to generate a key.

I do see a slight lag if I attempt to quickly type a bunch of random characters. However, for normal typing speeds, I do not see any noticeable difference compared to entering this comment on GitHub.

If I had to hazard a guess as to why the `name` input has a slightly longer lag than the filter box, it would be due to repaints and layouts. Namely, the `name` input node is slightly higher up the DOM tree and, thus, updates to it may result in larger tree updates and layout update surface area. But this is only a guess.

From reviewing the code, I don't see anything obvious that can be done to mitigate any lag, so going to mark this as resolved.

Reply:
> On my MacBook, I am unable to hold down a key to generate repeats. Even within this comment box on GitHub, I have to press each time I want to generate a key.

@kgryte Ohhh yeah, that's right. 5 or 6 os versions ago when they added that "show accented/alt chars on key hold" feature I had to turn it off:

```
defaults write -g ApplePressAndHoldEnabled -bool false
```

Much nicer that way, and there are also some special options for increasing the repeat speed beyond what the system preferences UI allows.

Anyhow, this is something to be aware of, but isn't a showstopper, so I think it's fine for now